kolibri/content/api.py

```diff
@@ -38,17 +41,93 @@ def title_description_filter(self, queryset, value):
         data = descendants | siblings  # concatenates different querysets
         return data
 
     def filter_recommendations(self, queryset, value):
-        # return 25 random content nodes
-        pks = queryset.values_list('pk', flat=True).exclude(kind__in=['topic', ''])
-        count = min(pks.count(), 25)
-        return queryset.filter(pk__in=sample(list(pks), count))
+        """
+        Recommend content items specific for this user.
+
+        :param queryset: all content nodes for this channel
+        :param value: id of currently logged in user, or none if user is anonymous
+        :return: recommended content nodes for user, or empty queryset if user is anonymous
+        """
+
+        from kolibri.logger.models import ContentSessionLog
+
+        if ContentSessionLog.objects.count() < 50:
+            # return 25 random content nodes if not enough session logs
+            pks = queryset.values_list('pk', flat=True).exclude(kind__in=['topic', ''])
+            count = min(pks.count(), 25)
+            return queryset.filter(pk__in=sample(list(pks), count))
+
+        # if user is anonymous, only give them the most popular content nodes
+        if value is None:
+            recently_viewed = queryset.objects.none()
+        else:
+            if self.data['channel']:  # filter by channel if available
+                user_session_logs = ContentSessionLog.objects.filter(user=value, channel_id=self.data['channel'])
+            else:
+                user_session_logs = ContentSessionLog.objects.filter(user=value)
+
+            # get the most recently viewed, but not finished, content nodes
+            content_ids = user_session_logs.exclude(progress=1).order_by('end_timestamp').values_list('content_id', flat=True).distinct()
+            recently_viewed = queryset.filter(content_id__in=list(content_ids[:10]))
+
+        return recently_viewed
+
+    def filter_next_steps(self, queryset, value):
+        """
+        Recommend uncompleted content, content that has user completed content as a prerequisite.
+
+        :param queryset: all content nodes for this channel
+        :param value: id of currently logged in user, or none if user is anonymous
+        :return: uncompleted content nodes, or empty queryset if user is anonymous
+        """
+        from kolibri.logger.models import ContentSummaryLog
+
+        # if user is anonymous, don't return any nodes
+        if value is None:
+            return queryset.objects.none()
+
+        if self.data['channel']:
+            summary_logs = ContentSummaryLog.objects.filter(user=value, channel_id=self.data['channel'])
+        else:
+            summary_logs = ContentSummaryLog.objects.filter(user=value)
+
+        content_ids = summary_logs.exclude(progress=1).values_list('content_id', flat=True)
+        unfinished_nodes = queryset.filter(content_id__in=list(content_ids[:10]))
+
+        return unfinished_nodes
+
+    def filter_popular(self, queryset, value):
```

Comment:
@jamalex I imagine this is something that might be worth caching to reduce database hits? Is there a good way of doing that?